lib/src/state_machine_controller.dart

```diff
@@ -1,17 +1,51 @@
+import 'package:flutter/foundation.dart';
+import 'package:rive/src/core/core.dart';
+import 'package:rive/src/generated/animation/state_machine_bool_base.dart';
+import 'package:rive/src/generated/animation/state_machine_number_base.dart';
+import 'package:rive/src/generated/animation/state_machine_trigger_base.dart';
 import 'package:rive/src/rive_core/animation/state_machine.dart';
+import 'package:rive/src/rive_core/animation/state_machine_bool.dart';
+import 'package:rive/src/rive_core/animation/state_machine_input.dart' as core;
+import 'package:rive/src/rive_core/animation/state_machine_number.dart';
+import 'package:rive/src/rive_core/animation/state_machine_trigger.dart';
 import 'package:rive/src/rive_core/artboard.dart';
 import 'package:rive/src/rive_core/state_machine_controller.dart' as core;
 
-class StateMachineInput<T> {
-  final int id;
+/// [StateMachine]s supports three input types. The StateMachine mostly
+/// abstracts types by allowing the programmer to query for an input of a
+/// specific Dart backing type, mapping it to the correct StateMachine type.
+/// This is the most flexible API to use to check if a type with a given name
+/// exists. However, if you need to iterate inputs and query their types, this
+/// enum is exposed for convenience.
+enum SMIType { number, boolean, trigger }
+
+/// StateMachine Instance Input. This is the abstraction of an instanced input
+/// from the [StateMachine]. Whenever a [StateMachineController] is created, the
+/// list of inputs in the corresponding [StateMachine] is wrapped into a set of
+/// [SMIInput] objects that ensure inputs are initialized to design-time values.
+/// The implementation can now change these values freely as they are decoupled
+/// from the backing [StateMachine] and can safely be re-instanced by another
+/// controller later.
+class SMIInput<T> {
```

Comment:
I am wondering, can / should this be an abstract class? Or is this instantiated somewhere?

```suggestion
abstract class SMIInput<T> {
```

Reply:
That's a great point, it should totally be abstract.